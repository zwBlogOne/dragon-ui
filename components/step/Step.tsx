import React, { Component, ReactElement } from 'react';
import classnames from 'classnames';
import Events from '../utils/events';
import StepItem from './StepItem';
import PropsType from './PropsType';

class Step extends Component<PropsType, any> {
  static defaultProps = {
    prefixCls: 'ui-step',
    current: 1,
  };

  static Item;

  private unmounted;

  constructor(props) {
    super(props);
    this.unmounted = false;
    this.state = {
      itemWidth: '100%',
    };
  }

  componentDidMount() {
    this.unmounted = true;
    this.handleUpdate();
    this.bindHandlers();
  }

  componentWillUnmount() {
    this.unmounted = false;
    this.unbindHandlers();
  }

  handleUpdate() {
    if (!this.unmounted) {
      return;
    }

    const num = React.Children.count(this.props.children);
    const itemWidth = `${100 / num}%`;

    this.setState({ itemWidth });
  }

  bindHandlers() {
    Events.on(window, 'resize', _ => this.handleUpdate());
  }

  unbindHandlers() {
    Events.off(window, 'resize', _ => this.handleUpdate());
  }

  render() {
    const { props } = this;
    const { className, current, style, prefixCls } = props;

    const cls = classnames({
      [prefixCls!]: true,
      [className!]: !!className,
    });

    const children = React.Children.map(props.children, (item, index) => (
      <StepItem
        {...(item as ReactElement<any>).props}
        isFinished={index + 1 < current}
        isProcess={index + 1 === Number(current)}
        index={index + 1}
        style={{ width: `${this.state.itemWidth}` }}
      />
    ));

    return (
      <div className={cls} style={style}>
        {children}
      </div>
    );
  }
}

export default Step;
