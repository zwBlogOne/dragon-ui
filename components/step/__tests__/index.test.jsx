import React from 'react';
import { render, mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import Step from '../index';

describe('Step', () => {
  it('renders normal Step correctly', () => {
    const wrapper = render(
      <div>
        <Step current={3}>
          <Step.Item>投保单基本信息</Step.Item>
          <Step.Item>投保单位录入</Step.Item>
          <Step.Item>产品选择</Step.Item>
          <Step.Item>总单险种定义</Step.Item>
          <Step.Item>计划创建</Step.Item>
          <Step.Item>被保人清单导入</Step.Item>
          <Step.Item>录入完成</Step.Item>
        </Step>
      </div>
    );

    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('renders StepItem correctly', () => {
    const wrapper = mount(
      <div>
        <Step current={3}>
          <Step.Item>投保单基本信息</Step.Item>
          <Step.Item>投保单位录入</Step.Item>
          <Step.Item>产品选择</Step.Item>
          <Step.Item>总单险种定义</Step.Item>
          <Step.Item>计划创建</Step.Item>
        </Step>
      </div>
    );

    expect(wrapper.find('.ui-step-item')).toHaveLength(5);
    expect(wrapper.find('.ui-step-item').at(2).hasClass('ui-step-item-process')).toBeTruthy();
  });

  it('computes itemWidth correctly', () => {
    const wrapper = mount(
      <Step current={3}>
        <Step.Item>投保单基本信息</Step.Item>
        <Step.Item>投保单位录入</Step.Item>
        <Step.Item>产品选择</Step.Item>
        <Step.Item>总单险种定义</Step.Item>
        <Step.Item>计划创建</Step.Item>
      </Step>
    );

    expect(wrapper.state('itemWidth')).toEqual('20%');
  });
});

