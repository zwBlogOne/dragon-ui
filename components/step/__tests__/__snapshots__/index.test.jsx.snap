// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Step renders normal Step correctly 1`] = `
<div>
  <div
    class="ui-step"
  >
    <div
      class="ui-step-item ui-step-item-finish"
      style="width:100%"
    >
      <div
        class="ui-step-item-line"
      />
      <div
        class="ui-step-item-header"
      >
        <div
          class="ui-step-item-header-inner"
        >
          <span>
            <i
              class="ui-icon ui-icon-right theme-default"
            />
          </span>
        </div>
      </div>
      <div
        class="ui-step-item-body"
      >
        投保单基本信息
      </div>
    </div>
    <div
      class="ui-step-item ui-step-item-finish"
      style="width:100%"
    >
      <div
        class="ui-step-item-line"
      />
      <div
        class="ui-step-item-header"
      >
        <div
          class="ui-step-item-header-inner"
        >
          <span>
            <i
              class="ui-icon ui-icon-right theme-default"
            />
          </span>
        </div>
      </div>
      <div
        class="ui-step-item-body"
      >
        投保单位录入
      </div>
    </div>
    <div
      class="ui-step-item ui-step-item-process"
      style="width:100%"
    >
      <div
        class="ui-step-item-line"
      />
      <div
        class="ui-step-item-header"
      >
        <div
          class="ui-step-item-header-inner"
        >
          <span>
            3
          </span>
        </div>
      </div>
      <div
        class="ui-step-item-body"
      >
        产品选择
      </div>
    </div>
    <div
      class="ui-step-item"
      style="width:100%"
    >
      <div
        class="ui-step-item-line"
      />
      <div
        class="ui-step-item-header"
      >
        <div
          class="ui-step-item-header-inner"
        >
          <span>
            4
          </span>
        </div>
      </div>
      <div
        class="ui-step-item-body"
      >
        总单险种定义
      </div>
    </div>
    <div
      class="ui-step-item"
      style="width:100%"
    >
      <div
        class="ui-step-item-line"
      />
      <div
        class="ui-step-item-header"
      >
        <div
          class="ui-step-item-header-inner"
        >
          <span>
            5
          </span>
        </div>
      </div>
      <div
        class="ui-step-item-body"
      >
        计划创建
      </div>
    </div>
    <div
      class="ui-step-item"
      style="width:100%"
    >
      <div
        class="ui-step-item-line"
      />
      <div
        class="ui-step-item-header"
      >
        <div
          class="ui-step-item-header-inner"
        >
          <span>
            6
          </span>
        </div>
      </div>
      <div
        class="ui-step-item-body"
      >
        被保人清单导入
      </div>
    </div>
    <div
      class="ui-step-item"
      style="width:100%"
    >
      <div
        class="ui-step-item-line"
      />
      <div
        class="ui-step-item-header"
      >
        <div
          class="ui-step-item-header-inner"
        >
          <span>
            7
          </span>
        </div>
      </div>
      <div
        class="ui-step-item-body"
      >
        录入完成
      </div>
    </div>
  </div>
</div>
`;
